### ISSUE TITLE
CUPS Fails to Detect Printer After Power Off

### ISSUE THREAD

**Issue by 0elek** [anujdatar/cups-docker#15]
#### Currently, when the printer is turned off, cups is unable to detect it when I turn it back on.

It works fine when i restart the docker container, so I think its having trouble finding the /dev/usb/[], since it changes when the printer is reconnected container host's OS.

Not sure that I set up something wrong, or its an issue with the container.

docker-compose.yml: 
```version: "3"
services:
    cups:
        image: anujdatar/cups
        container_name: cups
        restart: always
        ports:
            - "631:631"
        devices:
            - /dev/bus:/dev/bus
            - /dev/usb:/dev/usb
        environment:
            - CUPSADMIN=admin
            - CUPSPASSWORD=password
            - TZ="Europe/Warsaw"
        volumes:
            - ./data:/etc/cups
```

data/printers.conf: 
```# Printer configuration file for CUPS v2.4.2
# Written by cupsd
# DO NOT EDIT THIS FILE WHEN CUPSD IS RUNNING
NextPrinterId 2
<Printer EPSON_L310_Series>
PrinterId 1
UUID urn:uuid:aa358d43-05a5-31db-6324-f9e9d4e078b6
Info EPSON L310 Series
Location
MakeModel Epson L310 - CUPS+Gutenprint v5.3.4
DeviceURI usb://EPSON/L310%20Series?serial=56484D4B3032353618
State Idle
StateTime 1712484122
ConfigTime 1711806586
Type 8425484
Accepting Yes
Shared Yes
JobSheets none none
QuotaPeriod 0
PageLimit 0
KLimit 0
OpPolicy default
ErrorPolicy retry-job
</Printer>
```

**Comment by anujdatar** [comment]
Hi,

Sorry about that. I've been meaning to reply, but I haven't had time to really look into it. But you are correct, I have also experienced this in the past.

The container doesn't always recognise the printer, after the printer is turned OFF then ON. I had tried to avoid this by passing the entire `/dev/usb` device, instead of a specific port. But it doesn't always work as intended.

I typically have a push-button connected to my RPi's gpio that just restarts this container.

I'll try to take a look at the issue in detail and see what can be done. Can't give you a timeline, but I'll definitely take a look at it. It's something I need to fix anyway. Thanks for reminding me.

**Comment by mx3coding** [comment]
This is happening with me too.

It is possible to add an workaround, where the printer USB events are listened and if an "on/off" event occur then restart the docker-CUPS service?

**Comment by anujdatar** [comment]
This could work, but then, if the a printer's power is turned off, and it does not come back on, you could potentially get stuck in a restart-loop. Plus I am not sure how you would run a daemon like that unless you just run a checking script at regular intervals, systemd is typically not used in containers iirc.

**Comment by mx3coding** [comment]
You are right, this could lead to other issues...

About the restart-loop, my only idea is to create a variable (linked to a checkbox) where you need to be sure if you want to enable it or not, knowing the risks.

**Comment by 0elek** [comment]
i think first we should find out why does to container loses the usb device, cups on bare metal can handle power state change

**Comment by mx3coding** [comment]
We managed to put a udev rule to monitor USB changes and restart the container.
It not the solution for the "why the docker/container loses the connectivity to the device", but works for our needs.

If someone needs it, I detailed the solution here:
https://superuser.com/q/1849642/853051

**Comment by anujdatar** [comment]
Yeah, that was my first thought too, but that's a solution outside the container. So I didn't want to make that the go to method. But yeah, that's a fix for now.

I have an idea I want to explore within the container, for restarting the cups service. I will try it tomorrow and keep you posted.

**Comment by mx3coding** [comment]
Alright, I will be watching your project.

I have zero to none experience with docker, but if I learn something, I will share here.

**Comment by 0elek** [comment]
[this looks helpful](https://marc.merlins.org/perso/linux/post_2018-12-20_Accessing-USB-Devices-In-Docker-_ttyUSB0_-dev-bus-usb-_-for-fastboot_-adb_-without-using-privileged.html) but i think we need to run this outside of the container,

Also `privileged: true` seems to be able to forward the usb after on/off but the id changes,
so adding a udev rule(inside the container) should work then?
But for that we need somehow know some information like manufacturer and device id.
which we can get from the printers.conf.

this looks like a way if somebody wants to fix it.

**Comment by as4ajaysoni** [comment]
To handle automatic Docker container restarts when a USB printer is connected, I’ve written a udev-based installation script. It sets up the necessary rule and restart logic. 

The script is available here: 

https://github.com/as4ajaysoni/homelab/blob/main/scripts/install-printer-udev-rules.sh

**Comment by sacrefizz** [comment]
**Root cause + a fix that removes the need to restart the container** (re: printer off → on)

I hit this too and traced it. The culprit is `devices:` / `--device`. Docker resolves each device at **container-create time** and `mknod`s a *copy* of the node into the container's private `/dev` — a static snapshot. When the printer is powered off/on it re-enumerates and the host creates a **new** node (e.g. `/dev/bus/usb/001/007` → `001/008`); the container still holds the old node, now pointing at nothing, so the CUPS backend hangs on *"waiting for printer to become available"*. `docker restart` "fixes" it only because it re-snapshots `/dev`.

The fix is to give the container a **live view** of the host's `devtmpfs` instead of a snapshot — bind-mount `/dev/bus/usb` as a **volume** (not a device) so new nodes appear instantly, plus a wildcard cgroup rule so the re-enumerated node stays openable.

**docker-compose:**
```yaml
    # remove:  devices: [ /dev/bus/usb ]
    device_cgroup_rules:
      - 'c 189:* rmw'                 # usbfs — the CUPS libusb backend does all I/O here
    volumes:
      - ./data:/etc/cups
      - /dev/bus/usb:/dev/bus/usb     # live devtmpfs view -> re-enumeration visible, no restart
```

**docker run:**
```sh
docker run -d -p 631:631 \
  -v /dev/bus/usb:/dev/bus/usb \
  --device-cgroup-rule='c 189:* rmw' \
  --name cups anujdatar/cups
```

Why it works: creating a hotplug node is a filesystem operation on the host `devtmpfs`, **not** a mount event, so a bind-mount sees it immediately (mount-propagation mode is irrelevant). The old `devices:` entry was doing two jobs — exposing the node *and* granting the cgroup device permission — so once you drop it you must add `device_cgroup_rules` back or `open()` returns `EPERM`. `189` is the usbfs major; the **wildcard minor** is what survives re-enumeration (a fixed minor goes stale exactly like `devices:` did). `/dev/usb/lp*` isn't needed — the `usb://…` backend is libusb-based and detaches `usblp` via an ioctl through usbfs.

Tested end-to-end (Docker 27 / cgroup v2, kernel 6.12) with a Samsung M2020: powered the printer off, it came back as a **new** USB device number, the container saw it **live**, and the queued job printed with **no container restart**. Cleaner than `privileged: true` (which also works but grants every device).

I'll open a PR updating the README with this.
